### PR TITLE
Save cycle

### LINKED ISSUE #209: [editors] Support Manual Save
Currently, the editor always saves automatically. This should be an optional feature. In addition, an explicit save command should be provided to save changed editors. We need to implement
- [x] a save command, that takes the current editor contents and sends it to the backend for saving it to the FileSystem service
- [x] a keybinding in the context of an active editor to trigger that command
- [x] a main menu entry in the context of an active editor to trigger that command
- [x] an indicator on the tab of an editor, to indicate whether the editor is ‘dirty’ (i.e. contains local changes)
- [x] synchronize not dirty editors with changes in the file system
- [x] ask the user to reload from disk or override with the changes in the editor on save if the file was modified in the meantime
- [x] hook up with close action, i.e. ask the user to save before closing dirty editors
- [x] add `Save All` command
- [x] add `editor.autoSave` and `editor.autoSaveDelay` preferences

### PR DESCRIPTION
**Pull request by akosyakov**
Fixes #209

### CODE REVIEW

File: packages/core/src/common/menu.ts [hunk 1]
```diff
@@ -46,6 +46,16 @@ export class MenuModelRegistry {
         return parent.addNode(actionNode);
     }
 
+    registerSubMenu(subMenuPath: string[], label: string): Disposable {
```
Comment (akosyakov) [inline packages/core/src/common/menu.ts]: I would like to replace `registerSubmenu(menuPath, id, label)` with `registerSubmenu(subMenuPath, label)`, it is more convenient for clients to work with menu paths rather than with ids

File: packages/editor/package.json [hunk 2]
```diff
@@ -4,8 +4,8 @@
   "description": "Theia - Editor Extension",
   "dependencies": {
     "@theia/core": "^0.2.0",
-    "@theia/preferences": "^0.2.0",
-    "@theia/workspace": "^0.2.0"
+    "@theia/preferences-api": "^0.2.0",
+    "@theia/filesystem": "^0.2.0"
```
Comment (akosyakov) [inline packages/editor/package.json]: I think `FileIconProvider` should be moved to `@theia/core` from `@theia/filesystem` and renamed `IconProvider`, otherwise extensions are forced to depend on the filesystem

File: packages/monaco/src/browser/monaco-editor-model.ts [hunk 3]
```diff
@@ -5,23 +5,25 @@
...
     readonly model: monaco.editor.IModel;
     readonly reason: TextDocumentSaveReason;
     waitUntil(thenable: Thenable<monaco.editor.IIdentifiedSingleEditOperation[]>): void;
 }
 
-export class MonacoEditorModel implements ITextEditorModel {
+export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
 
-    autoSave: boolean = true;
+    autoSave: 'on' | 'off' = 'on';
```
Comment (akosyakov) [inline packages/monaco/src/browser/monaco-editor-model.ts]: I would like to move autosave functionality to `@theia/core`, i.e. it will work based `Saveable` and applied to all widgets, not only to editors. Objections?

File: packages/editor/package.json (hunk 2 above)
Comment (svenefftinge) [reply packages/editor/package.json]: ok

File: packages/monaco/src/browser/monaco-editor-model.ts (hunk 3 above)
Comment (svenefftinge) [reply packages/monaco/src/browser/monaco-editor-model.ts]: Will it be mandatory for all editing widgets to support both modes?

File: packages/core/src/browser/common-frontend-contribution.ts [hunk 4]
```diff
@@ -17,19 +17,19 @@ import { MAINAREA_TABBAR_CONTEXT_MENU } from './shell';
 
 export namespace CommonMenus {
 
-    export const EDIT_MENU = "2_edit";
-    export const EDIT_MENU_UNDO_GROUP = "1_undo/redo";
-    export const EDIT_MENU_CUT_COPY_PASTE_GROUP = "2_cut/copy/paste";
-    export const EDIT_MENU_FIND_REPLACE_GROUP = "3_find/replace";
+    export const FILE = [MAIN_MENU_BAR, '1_file'];
+    export const FILE_NEW = [...FILE, '1_new'];
```
Comment (svenefftinge) [inline packages/core/src/browser/common-frontend-contribution.ts]: `new` and `open ...` are very specific to editor and workspace. Shouldn't the declaration stay there?

File: packages/core/src/browser/dialogs.ts [hunk 5]
```diff
@@ -149,7 +149,7 @@ export abstract class AbstractDialog<T> extends BaseWidget {
     }
 
     protected addCloseAction<K extends keyof HTMLElementEventMap>(element: HTMLElement, ...additionalEventTypes: K[]): void {
-        this.addKeyListener(element, Key.ESCAPE, () => this.close(), ...additionalEventTypes);
+        this.addKeyListener(element, Key.ENTER, () => this.close(), ...additionalEventTypes);
```
Comment (svenefftinge) [inline packages/core/src/browser/dialogs.ts]: why did you change this to enter?

File: packages/monaco/src/browser/monaco-editor-model.ts (hunk 3 above)
Comment (akosyakov) [reply packages/monaco/src/browser/monaco-editor-model.ts]: I thought it would be supported automatically and configured with one common preference. But thinking again it is probably should be configurable per a widget type.

File: packages/core/src/browser/shell.ts [hunk 6]
```diff
@@ -64,7 +62,7 @@ export interface DockLayoutData extends DockPanel.ILayoutConfig {
     activeWidgets?: Widget[]
 }
 
-export const MAINAREA_TABBAR_CONTEXT_MENU = 'mainarea-tabbar-context-menu';
+export const MAINAREA_TABBAR_CONTEXT_MENU = ['mainarea-tabbar-context-menu'];
```
Comment (svenefftinge) [inline packages/core/src/browser/shell.ts]: It would be better if you turn it into an array on the use-side

File: packages/core/src/common/menu.ts (hunk 1 above)
Comment (svenefftinge) [reply packages/core/src/common/menu.ts]: ok

File: packages/editor/src/browser/editor-widget.ts [hunk 7]
```diff
@@ -6,19 +6,23 @@
...
     constructor(
         readonly editor: TextEditor,
-        readonly selectionService: SelectionService
+        protected readonly selectionService: SelectionService
     ) {
         super(editor);
         this.toDispose.push(this.editor);
     }
 
+    get saveable(): TextEditorDocument {
```
Comment (svenefftinge) [inline packages/editor/src/browser/editor-widget.ts]: should the return type not only be `Saveable`. We already expose the document through `editor.document`.

File: packages/filesystem/src/common/filesystem.ts [hunk 8]
```diff
@@ -102,6 +102,16 @@ export interface FileSystem extends Disposable {
 
 }
 
+export interface FileSystemClient {
+
+    /**
+     * Tests whether the given file can be overwrite
```
Comment (svenefftinge) [inline packages/filesystem/src/common/filesystem.ts]: overwritten

File: packages/monaco/src/browser/monaco-editor-model.ts (hunk 3 above)
Comment (svenefftinge) [reply packages/monaco/src/browser/monaco-editor-model.ts]: Ah, you simply want trigger save when `onDirtyChanged` signals?

File: packages/core/src/browser/dialogs.ts (hunk 5 above)
Comment (akosyakov) [reply packages/core/src/browser/dialogs.ts]: it was a bug, close button should be triggered on `enter` when it is focused, not on `esc`

Comment (svenefftinge) [reply packages/core/src/browser/dialogs.ts]: ok

File: packages/editor/package.json (hunk 2 above)
Comment (akosyakov) [reply packages/editor/package.json]: An issue for it: https://github.com/theia-ide/theia/issues/801

File: packages/core/src/browser/shell.ts (hunk 6 above)
Comment (svenefftinge) [reply packages/core/src/browser/shell.ts]: Maybe we should declare a type alias `MenuPath` for `string[]`.
And then also change contextMenuRenderer to accept that type.

File: packages/monaco/src/browser/monaco-editor-model.ts (hunk 3 above)
Comment (akosyakov) [reply packages/monaco/src/browser/monaco-editor-model.ts]: We've discussed that it will require moving prefereces-api to core. I would like to make it as the separate PR.

File: packages/core/src/browser/common-frontend-contribution.ts (hunk 4 above)
Comment (akosyakov) [reply packages/core/src/browser/common-frontend-contribution.ts]: ok

File: packages/core/src/browser/shell.ts (hunk 6 above)
Comment (akosyakov) [reply packages/core/src/browser/shell.ts]: ok

File: packages/core/src/browser/common-frontend-contribution.ts (hunk 4 above)
Comment (akosyakov) [reply packages/core/src/browser/common-frontend-contribution.ts]: @svenefftinge With open is easy, but new is used into the terminal extension. Could we move `new terminal` to `view` menu?

**Comment by akosyakov** [comment]
@svenefftinge I've addressed all comments

**Comment by svenefftinge** [comment]
If I try to save on a non-dirty editor, the browser-save as dialog pops up. That should no longer happen in any case. Simply leaving save command enabled always would be fine for me.

File: packages/core/src/browser/common-frontend-contribution.ts (hunk 4 above)
Comment (kittaakos) [reply packages/core/src/browser/common-frontend-contribution.ts]: +1 Just like in VS Code.

File: packages/core/src/browser/saveable.ts [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    export function is(arg: any): arg is Saveable {
+        return !!arg && ('dirty' in arg) && ('onDirtyChanged' in arg);
+    }
+    export function get(arg: any): Saveable | undefined {
+        if (is(arg)) {
+            return arg;
+        }
+        if (isSource(arg)) {
+            return arg.saveable;
+        }
```
Comment (kittaakos) [inline packages/core/src/browser/saveable.ts]: Can we have explicit `return undefined`? Same below in `getDirty`.

Comment (akosyakov) [reply packages/core/src/browser/saveable.ts]: I think the type system is capable to catch it, if you change a return type to `Saveable` it will be an error

File: packages/core/src/browser/saveable.ts [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+    protected onAfterAttach(msg: Message): void {
+        super.onAfterAttach(msg);
+        this.addKeyListener(this.dontSaveButton, Key.ENTER, () => {
+            this.shouldSave = false;
+            this.accept();
+        }, 'click');
+    }
+
+    protected shouldSave = true;
```
Comment (kittaakos) [inline packages/core/src/browser/saveable.ts]: Please move this up, next to the `dontSaveButton`. It would be better to see them together.

Comment (akosyakov) [reply packages/core/src/browser/saveable.ts]: why?

File: packages/editor/src/browser/editor-keybinding.ts [hunk 11]
```diff
@@ -31,21 +30,6 @@ export class EditorKeybindingContribution implements KeybindingContribution {
...
-            },
-            {
-                commandId: 'editor.close.all',
-                context: this.editorKeybindingContext,
-                keyCode: KeyCode.createKeyCode({ first: Key.KEY_W, modifiers: [Modifier.M2, Modifier.M3] })
-            }
-        ].forEach(binding => {
-            registry.registerKeyBinding(binding);
-        });
-
```
Comment (kittaakos) [inline packages/editor/src/browser/editor-keybinding.ts]: Is there a good reason to keep this?

Comment (akosyakov) [reply packages/editor/src/browser/editor-keybinding.ts]: just don't want to remove today to add it tomorrow

File: packages/monaco/src/browser/monaco-editor-model.ts [hunk 12]
```diff
@@ -33,29 +35,79 @@ export class MonacoEditorModel implements ITextEditorModel {
...
+    protected _dirty = false;
+    get dirty(): boolean {
+        return this._dirty;
+    }
+    protected setDirty(dirty: boolean): void {
+        this._dirty = dirty;
+        this.onDirtyChangedEmitter.fire(undefined);
+    }
+
+    protected readonly onDirtyChangedEmitter = new Emitter<void>();
```
Comment (kittaakos) [inline packages/monaco/src/browser/monaco-editor-model.ts]: Please pull these up if you do not have a good reason to keep them there.: `onDirtyChangedEmitter`, `_dirty` and `synchronizing`. Thanks!

Comment (akosyakov) [reply packages/monaco/src/browser/monaco-editor-model.ts]: They belong to the part where they used. Why is it important to declare everything in one place?

File: packages/monaco/src/browser/monaco-editor-model.ts [hunk 13]
```diff
@@ -86,8 +138,32 @@ export class MonacoEditorModel implements ITextEditorModel {
...
 
+    protected synchronizing = false;
+    async sync(): Promise<void> {
+        if (!this._dirty) {
+            const newText = await this.resource.readContents();
+            if (!this._dirty) {
+                const value = this.model.getValue();
+                if (value !== newText) {
+                    this.synchronizing = true;
+                    this.model.applyEdits([this.p2m.asTextEdit({
```
Comment (kittaakos) [inline packages/monaco/src/browser/monaco-editor-model.ts]: What do you think, wouldn't it make sense to wrap into a try/catch/finally? So it will not get stuck in synchronizing state if an error happens.

Comment (akosyakov) [reply packages/monaco/src/browser/monaco-editor-model.ts]: good point

**Comment by akosyakov** [comment]
@hexa00 please review

**Comment by hexa00** [comment]
@akosyakov doing that now

**Comment by hexa00** [comment]
Ctrl-Alt-S for save all is also not working..

**Comment by hexa00** [comment]
That's Ctrl-s and Ctrl-Alt-s sorry just so there's no confusion

**Comment by akosyakov** [comment]
> If a file as unsaved changes and the filesytem file changes.. there's no warning about that to the user until he tries to save.

i think va code and eclipse behaves the same

**Comment by hexa00** [comment]
>i think va code and eclipse behaves the same

Just tested that seems like you're right, I though emacs did something however but it seems like it isn't. so functionally no problems :)

I'm checking the code now

**Comment by akosyakov** [comment]
@hexa00 I fixed couple bugs in the Monaco editors caught by the new version. They improved the type system a bit.
